edk2toollib/acpi/ivrs_parser.py

```diff
@@ -0,0 +1,643 @@
+##
+# Copyright (C) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+# Python script that converts a raw IVRS table into a struct
+##
+
+import os
+import sys
+import struct
+import xml.etree.ElementTree as ET
+
+IVRSParserVersion = '1.00'
+
+
+class IVRS_TABLE(object):
+    # Header Lengths
+    IVRSHeaderLength = 48
+    IVHDHeader10HLength = 24
+    IVHDHeaderExLength = 40
+    DeviceTableEntryLengthShort = 4
+    DeviceTableEntryLengthLong = 8
+    DeviceTableEntryLengthVar = 22
+
+    def __init__(self, data):
```

Comment:
For each of the structures like this, let's standardize on this interface:
Encode(): Object -> Bytes string
Decode(): Bytes string -> Object fields

Eliminate the __str__ functions and change them to DumpInfo(). DumpInfo() should be patterned like https://github.com/tianocore/edk2-pytool-library/blob/83ea544be8d4a9225fb2c4caac8fc264558df2d0/edk2toollib/uefi/fmp_capsule_header.py#L279

Reply:
Also, in the __init__ function, let's make the "data" field optional and just invoke the Decode() function if it's provided.

Reply:
Encode and Decode methods are added to IVHD, IVMD and DTE classes to replace toBytes and other related parsing methods.